src/Lean/Server/Snapshots.lean

```diff
@@ -78,26 +92,32 @@ partial def parseAhead (contents : String) (snap : Snapshot) : IO (Array Syntax)
       else
         go inputCtx pmctx cmdParserState (stxs.push cmdStx)
 
-/-- Compiles the next command occurring after the given snapshot.
-If there is no next command (file ended), returns messages produced
-through the file. -/
+/-- Compiles the next command occurring after the given snapshot. If there is no next command
+(file ended), `Snapshot.isAtEnd` will hold of the return value. -/
 -- NOTE: This code is really very similar to Elab.Frontend. But generalizing it
 -- over "store snapshots"/"don't store snapshots" would likely result in confusing
 -- isServer? conditionals and not be worth it due to how short it is.
-def compileNextCmd (contents : String) (snap : Snapshot) : IO (Sum Snapshot MessageLog) := do
-  let inputCtx := Parser.mkInputContext contents "<input>"
+def compileNextCmd (text : FileMap) (snap : Snapshot) : IO Snapshot := do
+  let inputCtx := Parser.mkInputContext text.source "<input>"
   let cmdState := snap.cmdState
   let scope := cmdState.scopes.head!
   let pmctx := { env := cmdState.env, options := scope.opts, currNamespace := scope.currNamespace, openDecls := scope.openDecls }
   let (cmdStx, cmdParserState, msgLog) :=
     Parser.parseCommand inputCtx pmctx snap.mpState snap.msgLog
```

Comment:
Yeah, that would get rid of computing the `MessageLog` diff. But we do want `Snapshot.interactiveDiags` to contain every diagnostic up to a point so that we can send them easily without folding over all snapshots. So we'd have a `MessageLog` with messages just for this snapshot but an interactive diagnostics log for everything, which is perhaps a bit weird. What do you think?

Reply:
I think that's fine, we can call it `allInteractiveDiags` or something similar. Not sure folding would even be a performance issue, concatenating mostly-empty arrays should be quite fast even with hundreds of snapshots.

Reply:
I tried this out now, but it is still clunky. The reason is that the `MessageLog` is part of `Command.State` which is designed for a "rolling" usage. So we'd have to leave most of the command state as-is but clear out just the message log. I would leave it as is, but added a comment clarifying it.